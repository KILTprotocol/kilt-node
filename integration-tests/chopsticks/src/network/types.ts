import type { setupContext } from '@acala-network/chopsticks-testing'

export type Config = Awaited<ReturnType<typeof setupContext>>
