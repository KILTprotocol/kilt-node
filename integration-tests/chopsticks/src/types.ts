export interface AssetSwitchSupplyParameters {
	circulatingSupply: bigint
	sovereignSupply: bigint
	totalSupply: bigint
}
