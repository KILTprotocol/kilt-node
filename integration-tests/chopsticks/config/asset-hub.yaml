endpoint:
  - wss://asset-hub-rococo-rpc.dwellir.com
mock-signature-host: true
runtime-log-level: 5
port: 50001
db: db/asset-hub.sqlite

import-storage:
  System:
    Account:
      -
        -
          - 5Go9qhpqq2EzAZ1SKkkx6bTjnSoTnHEB8FHCnieNS57TimEF
        - providers: 1
          data:
            free: '100000000000000000000000'
