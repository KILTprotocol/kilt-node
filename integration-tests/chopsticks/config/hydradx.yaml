endpoint:
 - wss://hydradx-rpc.dwellir.com
 - wss://rpc.hydradx.cloud
db: ./db/hydradx.db.sqlite
port: 50002
runtime-log-level: 5

import-storage:
 TechnicalCommittee:
  Members:
   - 4rsmbFBYpVmWE2LHRsSZKxf22a8cbJLxMZqvVGkGkSDmDBcr
 Council:
  Members:
   - 4rsmbFBYpVmWE2LHRsSZKxf22a8cbJLxMZqvVGkGkSDmDBcr
 System:
  Account:
   -
    -
     - 4rsmbFBYpVmWE2LHRsSZKxf22a8cbJLxMZqvVGkGkSDmDBcr
    - providers: 1
      data:
       free: "1000000000000000000"
   -
    -
     - 4seWojfEHrk5YKPahdErazQ3CWEHZYi6NV4gKz5AaejWbRPJ
    - providers: 1
      data:
       free: "1000000000000000000"
 AssetRegistry:
  Assets:
   -
    -
     - 28
    - name: KILT
      assetType: Token
      existentialDeposit: 21358393848783
      symbol: KILT
      decimals: 15
      location:
       parents: 1
       interior: X1(Parachain(2086))
      isSufficient: true
  AssetIds:
   -
    -
     - KILT
    - 28
  AssetLocations:
  -
   -
    - 28
   - parents: 1
     interior:
      X1:
       Parachain: 2086
  LocationAssets:
   -
    -
     - parents: 1
       interior:
        X1:
         Parachain: 2086
    - 28
 MultiTransactionPayment:
  AcceptedCurrencies:
   -
    -
     - 28
    - "72639897479709500000"

