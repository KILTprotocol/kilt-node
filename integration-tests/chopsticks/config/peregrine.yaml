endpoint:
  - wss://peregrine.kilt.io
mock-signature-host: true
db: db/peregrine.sqlite
runtime-log-level: 5
wasm-override: /home/antonio/Developer/kilt-node/target/debug/wbuild/peregrine-runtime/peregrine_runtime.compact.compressed.wasm
