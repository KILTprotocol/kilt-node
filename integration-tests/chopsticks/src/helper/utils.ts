import { ApiPromise, SubmittableResult } from '@polkadot/api'
import { Keyring } from '@polkadot/keyring'
import { u8aToHex } from '@polkadot/util'
import { decodeAddress } from '@polkadot/util-crypto'

import type { SubmittableExtrinsic } from '@polkadot/api/types'
import type { KeyringPair } from '@polkadot/keyring/types'
import type { Codec } from '@polkadot/types/types'
import type { ExpectStatic } from 'vitest'
import { EventRecord } from '@polkadot/types/interfaces'

/**
 * Keyring instance to generate key pairs with 38 ss58Format encoding.
 */
const keyring = new Keyring({ type: 'ed25519', ss58Format: 38 })

/**
 * Key pairs for Alice, Bob, and Charlie used for testing.
 */
export const keysAlice = keyring.addFromUri('//alice', undefined)
export const keysBob = keyring.addFromUri('//bob', undefined)
export const keysCharlie = keyring.addFromUri('//charlie', undefined)

/**
 * Converts a given value to a Number
 */
export function toNumber(value: string | undefined): number | undefined {
	if (value === undefined) {
		return undefined
	}
	return Number(value)
}

/**
 * Converts a given address to its hexadecimal representation.
 *
 * @param addr - The address to be converted. It should be a string representation of an address.
 */
export function hexAddress(addr: string) {
	return u8aToHex(decodeAddress(addr))
}

/**
 * Validates if the received balance is within the expected range considering a certain precision.
 *
 * @param previousBalance - The balance before the operation.
 * @param receivedBalance - The balance after the operation.
 * @param deltaBalance - The expected change in balance.
 * @param expect - The assertion function from the testing library.
 * @param precision - The precision of the balance validation. It must be a value between 0 and 100.
 *
 * The function calculates the expected balance by adding the deltaBalance to the previousBalance.
 * It then calculates the lower and upper bounds of the expected balance considering the precision.
 * The received balance is then checked if it falls within the range of the lower and upper bounds.
 *
 * If the precision is not between 0 and 100, the function throws an error.
 *
 * @throws {Error} If the precision is not between 0 and 100.
 */
export function validateBalanceWithPrecision(
	previousBalance: bigint,
	receivedBalance: bigint,
	deltaBalance: bigint,
	expect: ExpectStatic,
	precision: bigint
) {
	if (precision < 0n || precision > 100n) {
		throw new Error('Precision must be between 0 and 100')
	}

	const allowedError = 100n - precision
	const expectedBalance = previousBalance + deltaBalance

	const lowerBound = expectedBalance - (expectedBalance * allowedError) / 100n
	const upperBound = expectedBalance + (expectedBalance * allowedError) / 100n

	expect(receivedBalance).toBeGreaterThanOrEqual(lowerBound)
	expect(receivedBalance).toBeLessThanOrEqual(upperBound)
}

/**
 * Fetches the paid fees for the executed XCM message. Is only working on the sender chain.
 */
export async function getPaidXcmFees(api: ApiPromise, eventRecords: Codec[]): Promise<bigint> {
	const polkadotFees = eventRecords.find((eventRecord) =>
		api.events.polkadotXcm.FeesPaid.is((eventRecord as EventRecord).event)
	)
	if (!polkadotFees) {
		throw new Error('FeesPaid event not found')
	}

	// eslint-disable-next-line @typescript-eslint/no-explicit-any
	return (polkadotFees as any).event.data.fees[0].fun.asFungible.toBigInt()
}

/**
 * Calculates the transaction fees for a given transaction.
 */
export async function calculateTxFees(
	tx: SubmittableExtrinsic<'promise', SubmittableResult>,
	account: KeyringPair
): Promise<bigint> {
	const paymentInfo = await tx.paymentInfo(account)
	return paymentInfo.partialFee.toBigInt()
}

// Constants for the balances
export const KILT = BigInt(1e15)
export const DOT = BigInt(1e10)
export const HDX = BigInt(1e12)

export const initialBalanceKILT = 100n * KILT
export const initialBalanceDOT = 100n * DOT
export const initialBalanceHDX = 100n * HDX
