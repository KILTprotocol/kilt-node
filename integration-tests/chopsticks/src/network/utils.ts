export function getSiblingLocation(paraId: number) {
	return {
		parents: 1,
		interior: {
			X1: { Parachain: paraId },
		},
	}
}

export function getParentLocation() {
	return {
		parents: 1,
		interior: 'Here',
	}
}

export function getAccountLocationV2(addr: string) {
	return {
		V2: {
			parents: 0,
			interior: {
				X1: {
					AccountId32: {
						network: 'Any',
						id: addr,
					},
				},
			},
		},
	}
}

export function getAccountLocationV3(addr: string) {
	return {
		V3: {
			parents: 0,
			interior: {
				X1: {
					AccountId32: {
						id: addr,
					},
				},
			},
		},
	}
}

export function getNativeAssetIdLocation(amount: bigint) {
	return {
		id: { Concrete: { parents: 0, interior: 'Here' } },
		fun: { Fungible: amount },
	}
}
