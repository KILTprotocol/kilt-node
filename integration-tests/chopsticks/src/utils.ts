import { Keyring } from '@polkadot/keyring'

import { AssetSwitchSupplyParameters } from './types.js'

const keyring = new Keyring({ type: 'ed25519', ss58Format: 38 })

export const keysAlice = keyring.addFromUri('//alice', undefined, 'ed25519')
export const keysBob = keyring.addFromUri('//bob', undefined, 'ed25519')
export const keysCharlie = keyring.addFromUri('//charlie', undefined, 'ed25519')

export function toNumber(value: string | undefined): number | undefined {
	if (value === undefined) {
		return undefined
	}

	return Number(value)
}

export function getAssetSwitchParameters(
	totalSupply: bigint = BigInt('100000000000000000000'),
	ratioCirculating: bigint = BigInt(10)
): AssetSwitchSupplyParameters {
	const circulatingSupply = (totalSupply * ratioCirculating) / BigInt(100)
	const sovereignSupply = totalSupply - circulatingSupply
	return { circulatingSupply, sovereignSupply, totalSupply }
}

export const KILT = BigInt(1e15)
export const DOT = BigInt(1e10)
export const HDX = BigInt(1e12)
export const ROC = BigInt(1e12)

export const initialBalanceKILT = BigInt(100) * KILT
export const initialBalanceDOT = BigInt(100) * DOT
export const initialBalanceHDX = BigInt(100) * HDX
export const initialBalanceROC = BigInt(100) * ROC
