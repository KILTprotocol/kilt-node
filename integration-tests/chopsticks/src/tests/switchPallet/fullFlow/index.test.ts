import { describe, beforeEach, it, afterEach } from 'vitest'
import { sendTransaction, withExpect } from '@acala-network/chopsticks-testing'
import type { KeyringPair } from '@polkadot/keyring/types'

import { createBlock } from '../../../network/utils.js'
import { calculateTxFees, getPaidXcmFees, hexAddress } from '../../../helper/utils.js'
import { testCases } from './config.js'
import { Config } from '../../../network/types.js'
import { checkSwitchPalletInvariant } from '../index.js'
import { spinUpNetwork, tearDownNetwork } from '../../utils.js'

describe.each(testCases)(
	'Switch KILTs full flow',

	({ account, query, txContext, config, sovereignAccount }) => {
		let nativeContext: Config
		let foreignContext: Config
		let relayContext: Config
		let senderAccount: KeyringPair
		const { desc } = config

		// Create the network context
		beforeEach(async () => {
			const { parachainContexts, relayChainContext } = await spinUpNetwork(config)
			nativeContext = parachainContexts[0]
			foreignContext = parachainContexts[1]
			relayContext = relayChainContext
			senderAccount = account
		})

		// Shut down the network
		afterEach(async () => {
			await tearDownNetwork([nativeContext, foreignContext, relayContext])
		})

		it(desc, async ({ expect }) => {
			const { checkEvents, checkSystemEvents } = withExpect(expect)
			const { tx, balanceToTransfer, events } = txContext
			const foreignFundsBeforeTx = await query.foreign.nativeFunds(foreignContext, senderAccount.address)

			//action
			// 1. send foreign tokens from foreign chain to native chain
			const txSendForeignAsset = tx.foreign.transfer(
				foreignContext,
				hexAddress(senderAccount.address),
				balanceToTransfer.foreign.toString()
			)

			const events1 = await sendTransaction(txSendForeignAsset.signAsync(senderAccount))

			// process tx
			await createBlock(foreignContext)
			// process xcm message
			await createBlock(nativeContext)

			// check balance movement
			const txFees = await calculateTxFees(txSendForeignAsset, senderAccount)

			const foreignFundsAfterTx = await query.foreign.nativeFunds(foreignContext, senderAccount.address)

			const xcmFees = await getPaidXcmFees(await events1.events)

			expect(foreignFundsBeforeTx - balanceToTransfer.foreign - txFees - xcmFees).toBe(foreignFundsAfterTx)

			// check events
			await Promise.all(
				events.foreign.transfer.map((pallet) =>
					checkEvents(events1, pallet).toMatchSnapshot(
						`transfer foreign funds from foreign chain ${JSON.stringify(pallet)}`
					)
				)
			)

			await Promise.all(
				events.native.receive.foreign.map((pallet) =>
					checkSystemEvents(nativeContext, pallet).toMatchSnapshot(
						`receive foreign funds on native chain ${JSON.stringify(pallet)}`
					)
				)
			)

			// 2. send native tokens
			const nativeBalanceBeforeTx = await query.native.nativeFunds(nativeContext, senderAccount.address)

			// Send funds from native to foreign chainbalanceToTransferBackForeignparaId
			const txSendNativeAsset = tx.native.transfer(
				nativeContext,
				hexAddress(senderAccount.address),
				balanceToTransfer.native.toString()
			)

			const events2 = await sendTransaction(txSendNativeAsset.signAsync(senderAccount))
			// process tx
			await createBlock(nativeContext)
			// process xcm message
			await createBlock(foreignContext)

			// check balance movement

			const txFees2 = await calculateTxFees(txSendNativeAsset, senderAccount)

			const nativeBalanceAfterTx = await query.native.nativeFunds(nativeContext, senderAccount.address)

			expect(nativeBalanceBeforeTx - balanceToTransfer.native - txFees2).toBe(nativeBalanceAfterTx)

			// check events
			await Promise.all(
				events.native.transfer.map((pallet) =>
					checkEvents(events2, pallet).toMatchSnapshot(
						`Transfer native funds to foreign chain ${JSON.stringify(pallet)}`
					)
				)
			)

			await Promise.all(
				events.foreign.receive.native.map((pallet) =>
					checkSystemEvents(foreignContext, pallet).toMatchSnapshot(
						`Receive native funds on foreign chain ${JSON.stringify(pallet)}`
					)
				)
			)

			await checkSwitchPalletInvariant(
				expect,
				nativeContext,
				foreignContext,
				sovereignAccount,
				query.native.nativeFunds,
				query.foreign.foreignFunds
			)

			// 3. send native tokens back to sender chain.
			const balanceToTransferBack = balanceToTransfer.native / 2n

			const nativeBalanceForeignChainBeforeTx = await query.foreign.foreignFunds(
				foreignContext,
				hexAddress(senderAccount.address)
			)

			const signedTx3 = tx.native
				.withdraw(foreignContext, balanceToTransferBack.toString())
				.signAsync(senderAccount)

			const events3 = await sendTransaction(signedTx3)

			// process tx
			await createBlock(foreignContext)
			// process xcm message
			await createBlock(nativeContext)

			// check balance movement

			const nativeBalanceForeignChainAfterTx = await query.foreign.foreignFunds(
				foreignContext,
				senderAccount.address
			)

			expect(nativeBalanceForeignChainBeforeTx - balanceToTransferBack).toBe(nativeBalanceForeignChainAfterTx)

			// check events

			await Promise.all(
				events.foreign.withdraw.map((pallet) =>
					checkEvents(events3, pallet).toMatchSnapshot(
						`Withdraw native funds on foreign chain ${JSON.stringify(pallet)}`
					)
				)
			)

			await Promise.all(
				events.native.receive.native.map((pallet) =>
					checkSystemEvents(nativeContext, pallet).toMatchSnapshot(
						`Receive native funds on native chain ${JSON.stringify(pallet)}`
					)
				)
			)

			// finalize the switch. Create a another block to process the query xcm message
			await createBlock(nativeContext)
			await checkSwitchPalletInvariant(
				expect,
				nativeContext,
				foreignContext,
				sovereignAccount,
				query.native.nativeFunds,
				query.foreign.foreignFunds
			)

			// 4. send foreign token back

			const balanceToTransferBackForeign = balanceToTransfer.foreign / 10n

			const foreignBalanceBeforeTx = await query.native.foreignFunds(nativeContext, senderAccount.address)

			const signedTx4 = tx.foreign
				.withdraw(nativeContext, hexAddress(senderAccount.address), balanceToTransferBackForeign.toString())
				.signAsync(senderAccount)

			const events4 = await sendTransaction(signedTx4)

			// process tx
			await createBlock(nativeContext)

			// process xcm message
			await createBlock(foreignContext)

			// check balance movement
			const foreignBalanceAfterTx = await query.native.foreignFunds(nativeContext, senderAccount.address)
			expect(foreignBalanceBeforeTx - balanceToTransferBackForeign).toBe(foreignBalanceAfterTx)

			// check events

			await Promise.all(
				events.native.withdraw.map((pallet) =>
					checkEvents(events4, pallet).toMatchSnapshot(
						`Withdraw foreign funds on native chain ${JSON.stringify(pallet)}`
					)
				)
			)

			await Promise.all(
				events.foreign.receive.native.map((pallet) =>
					checkSystemEvents(foreignContext, pallet).toMatchSnapshot(
						`Receive foreign funds on foreign chain ${JSON.stringify(pallet)}`
					)
				)
			)
		})
	}
)
