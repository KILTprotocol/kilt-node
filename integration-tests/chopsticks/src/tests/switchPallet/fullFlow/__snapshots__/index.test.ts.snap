// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Switch KILTs full flow > V4 LIVE > Receive foreign funds on foreign chain {"section":"balances","method":"Burned"} 1`] = `
[
  {
    "data": {
      "amount": 1000000000,
      "who": "13cKp88mpGREFCq8KsJEFjpSBnjFuCNWq6bmD3js7fu4f66e",
    },
    "method": "Burned",
    "section": "balances",
  },
]
`;

exports[`Switch KILTs full flow > V4 LIVE > Receive foreign funds on foreign chain {"section":"balances","method":"Minted"} 1`] = `
[
  {
    "data": {
      "amount": "(rounded 960000000)",
      "who": "15jSz35ugoWTc61xHPoxEkHte4o7UanKCk1gx1dizA8yuNs8",
    },
    "method": "Minted",
    "section": "balances",
  },
]
`;

exports[`Switch KILTs full flow > V4 LIVE > Receive foreign funds on foreign chain {"section":"messageQueue","method":"Processed"} 1`] = `
[
  {
    "data": {
      "id": "(hash)",
      "origin": {
        "Sibling": "(rounded 2100)",
      },
      "success": true,
      "weightUsed": {
        "proofSize": "(rounded 7200)",
        "refTime": "(rounded 320000000)",
      },
    },
    "method": "Processed",
    "section": "messageQueue",
  },
]
`;

exports[`Switch KILTs full flow > V4 LIVE > Receive native funds on foreign chain {"section":"balances","method":"Burned"} 1`] = `
[
  {
    "data": {
      "amount": 5000000000,
      "who": "13cKp88mpGREFCq8KsJEFjpSBnjFuCNWq6bmD3js7fu4f66e",
    },
    "method": "Burned",
    "section": "balances",
  },
]
`;

exports[`Switch KILTs full flow > V4 LIVE > Receive native funds on foreign chain {"section":"balances","method":"Minted"} 1`] = `
[
  {
    "data": {
      "amount": "(rounded 300000000)",
      "who": "14xmwinmCEz6oRrFdczHKqHgWNMiCysE2KrA4jXXAAM1Eogk",
    },
    "method": "Minted",
    "section": "balances",
  },
  {
    "data": {
      "amount": "(rounded 4600000000)",
      "who": "15jSz35ugoWTc61xHPoxEkHte4o7UanKCk1gx1dizA8yuNs8",
    },
    "method": "Minted",
    "section": "balances",
  },
]
`;

exports[`Switch KILTs full flow > V4 LIVE > Receive native funds on foreign chain {"section":"messageQueue","method":"Processed"} 1`] = `
[
  {
    "data": {
      "id": "(hash)",
      "origin": {
        "Sibling": "(rounded 2100)",
      },
      "success": true,
      "weightUsed": {
        "proofSize": "(rounded 24000)",
        "refTime": "(rounded 1500000000)",
      },
    },
    "method": "Processed",
    "section": "messageQueue",
  },
]
`;

exports[`Switch KILTs full flow > V4 LIVE > Receive native funds on native chain {"section":"assetSwitchPool1","method":"RemoteToLocalSwitchExecuted"} 1`] = `
[
  {
    "data": {
      "amount": "(rounded 500000000000000)",
      "to": "4seWojfEHrk5YKPahdErazQ3CWEHZYi6NV4gKz5AaejWbRPJ",
    },
    "method": "RemoteToLocalSwitchExecuted",
    "section": "assetSwitchPool1",
  },
]
`;

exports[`Switch KILTs full flow > V4 LIVE > Transfer native funds to foreign chain {"section":"assetSwitchPool1","method":"LocalToRemoteSwitchExecuted"} 1`] = `
[
  {
    "data": {
      "amount": 1000000000000000,
      "from": "4seWojfEHrk5YKPahdErazQ3CWEHZYi6NV4gKz5AaejWbRPJ",
      "to": {
        "V4": {
          "interior": {
            "X1": [
              {
                "AccountId32": {
                  "id": "(hash)",
                  "network": null,
                },
              },
            ],
          },
          "parents": 0,
        },
      },
    },
    "method": "LocalToRemoteSwitchExecuted",
    "section": "assetSwitchPool1",
  },
]
`;

exports[`Switch KILTs full flow > V4 LIVE > Transfer native funds to foreign chain {"section":"fungibles","method":"Burned"} 1`] = `
[
  {
    "data": {
      "assetId": {
        "interior": "Here",
        "parents": 1,
      },
      "balance": 5000000000,
      "owner": "4seWojfEHrk5YKPahdErazQ3CWEHZYi6NV4gKz5AaejWbRPJ",
    },
    "method": "Burned",
    "section": "fungibles",
  },
]
`;

exports[`Switch KILTs full flow > V4 LIVE > Withdraw foreign funds on native chain "fungibles" 1`] = `
[
  {
    "data": {
      "assetId": {
        "interior": "Here",
        "parents": 1,
      },
      "balance": 1000000000,
      "owner": "4seWojfEHrk5YKPahdErazQ3CWEHZYi6NV4gKz5AaejWbRPJ",
    },
    "method": "Burned",
    "section": "fungibles",
  },
]
`;

exports[`Switch KILTs full flow > V4 LIVE > Withdraw foreign funds on native chain {"section":"polkadotXcm","method":"Sent"} 1`] = `
[
  {
    "data": {
      "destination": {
        "interior": {
          "X1": [
            {
              "Parachain": 1000,
            },
          ],
        },
        "parents": 1,
      },
      "message": [
        {
          "WithdrawAsset": [
            {
              "fun": {
                "Fungible": 1000000000,
              },
              "id": {
                "interior": "Here",
                "parents": 1,
              },
            },
          ],
        },
        "ClearOrigin",
        {
          "BuyExecution": {
            "fees": {
              "fun": {
                "Fungible": 1000000000,
              },
              "id": {
                "interior": "Here",
                "parents": 1,
              },
            },
            "weightLimit": "Unlimited",
          },
        },
        {
          "DepositAsset": {
            "assets": {
              "Wild": {
                "AllCounted": 1,
              },
            },
            "beneficiary": {
              "interior": {
                "X1": [
                  {
                    "AccountId32": {
                      "id": "(hash)",
                      "network": null,
                    },
                  },
                ],
              },
              "parents": 0,
            },
          },
        },
      ],
      "messageId": "(hash)",
      "origin": {
        "interior": {
          "X1": [
            {
              "AccountId32": {
                "id": "(hash)",
                "network": "Polkadot",
              },
            },
          ],
        },
        "parents": 0,
      },
    },
    "method": "Sent",
    "section": "polkadotXcm",
  },
]
`;

exports[`Switch KILTs full flow > V4 LIVE > Withdraw native funds on foreign chain {"section":"polkadotXcm","method":"Sent"} 1`] = `
[
  {
    "data": {
      "destination": {
        "interior": {
          "X1": [
            {
              "Parachain": "(rounded 2100)",
            },
          ],
        },
        "parents": 1,
      },
      "message": [
        {
          "ReserveAssetDeposited": [
            {
              "fun": {
                "Fungible": 500000000000000,
              },
              "id": {
                "interior": {
                  "X2": [
                    {
                      "GlobalConsensus": {
                        "Ethereum": {
                          "chainId": 1,
                        },
                      },
                    },
                    {
                      "AccountKey20": {
                        "key": "0x5d3d01fd6d2ad1169b17918eb4f153c6616288eb",
                        "network": null,
                      },
                    },
                  ],
                },
                "parents": 2,
              },
            },
          ],
        },
        "ClearOrigin",
        {
          "BuyExecution": {
            "fees": {
              "fun": {
                "Fungible": 500000000000000,
              },
              "id": {
                "interior": {
                  "X2": [
                    {
                      "GlobalConsensus": {
                        "Ethereum": {
                          "chainId": 1,
                        },
                      },
                    },
                    {
                      "AccountKey20": {
                        "key": "0x5d3d01fd6d2ad1169b17918eb4f153c6616288eb",
                        "network": null,
                      },
                    },
                  ],
                },
                "parents": 2,
              },
            },
            "weightLimit": "Unlimited",
          },
        },
        {
          "DepositAsset": {
            "assets": {
              "Wild": "All",
            },
            "beneficiary": {
              "interior": {
                "X1": [
                  {
                    "AccountId32": {
                      "id": "(hash)",
                      "network": null,
                    },
                  },
                ],
              },
              "parents": 0,
            },
          },
        },
      ],
      "messageId": "(hash)",
      "origin": {
        "interior": {
          "X1": [
            {
              "AccountId32": {
                "id": "(hash)",
                "network": "Polkadot",
              },
            },
          ],
        },
        "parents": 0,
      },
    },
    "method": "Sent",
    "section": "polkadotXcm",
  },
]
`;

exports[`Switch KILTs full flow > V4 LIVE > receive foreign funds on native chain {"section":"fungibles","method":"Issued"} 1`] = `
[
  {
    "data": {
      "amount": "(rounded 10000000000)",
      "assetId": {
        "interior": "Here",
        "parents": 1,
      },
      "owner": "4seWojfEHrk5YKPahdErazQ3CWEHZYi6NV4gKz5AaejWbRPJ",
    },
    "method": "Issued",
    "section": "fungibles",
  },
]
`;

exports[`Switch KILTs full flow > V4 LIVE > receive foreign funds on native chain {"section":"messageQueue","method":"Processed"} 1`] = `
[
  {
    "data": {
      "id": "(hash)",
      "origin": {
        "Sibling": 1000,
      },
      "success": true,
      "weightUsed": {
        "proofSize": 0,
        "refTime": 1000000000,
      },
    },
    "method": "Processed",
    "section": "messageQueue",
  },
]
`;

exports[`Switch KILTs full flow > V4 LIVE > transfer foreign funds from foreign chain {"section":"polkadotXcm","method":"Sent"} 1`] = `
[
  {
    "data": {
      "destination": {
        "interior": {
          "X1": [
            {
              "Parachain": "(rounded 2100)",
            },
          ],
        },
        "parents": 1,
      },
      "message": [
        {
          "ReserveAssetDeposited": [
            {
              "fun": {
                "Fungible": 10000000000,
              },
              "id": {
                "interior": "Here",
                "parents": 1,
              },
            },
          ],
        },
        "ClearOrigin",
        {
          "BuyExecution": {
            "fees": {
              "fun": {
                "Fungible": 10000000000,
              },
              "id": {
                "interior": "Here",
                "parents": 1,
              },
            },
            "weightLimit": "Unlimited",
          },
        },
        {
          "DepositAsset": {
            "assets": {
              "Wild": {
                "AllCounted": 1,
              },
            },
            "beneficiary": {
              "interior": {
                "X1": [
                  {
                    "AccountId32": {
                      "id": "(hash)",
                      "network": null,
                    },
                  },
                ],
              },
              "parents": 0,
            },
          },
        },
      ],
      "messageId": "(hash)",
      "origin": {
        "interior": {
          "X1": [
            {
              "AccountId32": {
                "id": "(hash)",
                "network": "Polkadot",
              },
            },
          ],
        },
        "parents": 0,
      },
    },
    "method": "Sent",
    "section": "polkadotXcm",
  },
]
`;
